### ISSUE TITLE
order gets changed while conversion.

### ISSUE THREAD

**Issue by sandext12** [nashwaan/xml-js#33]
Hi , 

I am trying to convert xml data to json and after some processing i want to convert back to xml. For testing purpose i am converting below data which is part of my xml file.

        <ProfAlign name="Layout">
		<PVI>0.77.906</PVI>
		<ParaCurve length="200.">247.13 91.807</ParaCurve>
		<PVI>1658.62622142 104.85033164</PVI>
       </ProfAlign>

xml2json ({compact: true, spaces: 4}) converts this data as below,

"ProfAlign": [
                                {
                                    "_attributes": {
                                        "name": "Layout"
                                    },
                                    "PVI": [
                                        {
                                            "_text": "0. 77.906"
                                        },
                                        {
                                            "_text": "1658.62622142 104.85033164"
                                        }
                                    ],
                                    "ParaCurve": {
                                        "_attributes": {
                                            "length": "200."
                                        },
                                        "_text": "247.13 91.807"
                                    }
                                }

And json2xml ({indentAttributes: false, spaces: '\t', fullTagEmptyElement: true, compact: true}) converts back to xml like below,

      <ProfAlign name="Layout">
		<PVI>0.77.906</PVI>
		<PVI>1658.62622142 104.85033164</PVI>
		<ParaCurve length="200.">247.13 91.807</ParaCurve>
      </ProfAlign>

As you can see, order is changed. Please suggest how to handle this.

**Comment by nashwaan** [comment]
To preserve the order, use `compact: false`.

Using `comact: true` will not preserve the order and this is how most other libraries are doing.
See [Compact Vs Non-Compact](https://github.com/nashwaan/xml-js#compact-vs-non-compact) for more details.
Actually, this is the main reason why I created this library: to preserve the order of the elements.

**Comment by sandext12** [comment]
Yes It worked, but made the json object very bulky. Thanks

**Comment by nashwaan** [comment]
Parsing of entities like `&quot;` and `&#xd'` are currently not supported.

I will try to add new option, say `parseEntities`, to support this in the coming few days.

**Comment by sandext12** [comment]
I have data in my xml like, (Please do not consider space after slash characters)

<collection name="arguments"/ >
<string name="connect_timeout"'></ string>

with fullTagEmptyElement: false =>
<collection name="arguments"/ >
<string name="connect_timeout"/ >

with fullTagEmptyElement: true =>
<collection name="arguments"/ ></ collectionProp>
<string name="connect_timeout"'></ string>

Can we handle these two scenarios individually ? please suggest.

**Comment by nashwaan** [comment]
Do you mean you want for certain elements to get output like `<a/>` and for other elements it should output `<b></b>`?

If this is what you want, then how does the converter knows that `a` elements should be shortened and `b` elements should have full closing tag?

**Comment by sandext12** [comment]
Yes this is what i want as my input is like this and i can not help here. Is it possible,  we can put some flag in converted json and use that flag to take appropriate action whether it should be <a/ > or <b></ b> ?

**Comment by nashwaan** [comment]
@sandext12 

Regarding encoding and decoding of html entities, it is actually [tricky and and there are many edge cases](https://stackoverflow.com/a/23831239/5619747). Therefore, I suggest to do this manually by utilizing [he](http://mths.be/he) library which was created to address this thoroughly. I could have include it in my library but it would increase its size and probably not needed for most users.

Regarding generation of `<a/>` and `<b></b>`, I am still not sure how to do this in a generic way. Can you provide an example of how would you pass the `options` for this requirement?

**Comment by nashwaan** [comment]
@sandext12
I have published version `1.6.1` of this library which can help you to decide which _empty elements_ can have full closing tags and which can have self closing tags. Just supply a function to `fullTagEmptyElementFn` option.

Example: 
```js
json2xml(json, {compact: true, fullTagEmptyElementFn: function(elementName) {
  return elementName === 'string'; // returning true forces to generate full closing tag 
}});
```

This way, you can achieve the desired output:
```xml
<collection name="arguments"/>
<string name="connect_timeout"></string>
```

For other functions, [see this](https://github.com/nashwaan/xml-js#options-for-custom-processing-functions).

**Comment by nashwaan** [comment]
Closed as no further comments received.
